File: Linked Lists/isListPalindrome.py
# Singly-linked lists are already defined with this interface:
# class ListNode(object):
#   def __init__(self, x):
#     self.value = x
#     self.next = None
#
def solution(l):
   
    middle = fast =l
    firstReverseHalf = list()
    while fast and fast.next:
        # get the middle
        middle = middle.next
        fast = fast.next.next
        
        # get the reverse of first half nodes
        temp = l.next
        l.next = firstReverseHalf
        firstReverseHalf = l
        l = temp
    
    if fast:
        middle = middle.next
    
    while middle:
        if middle.value != firstReverseHalf.value:
            return False
        middle = middle.next
        firstReverseHalf = firstReverseHalf.next
        
    return True
        

File: Linked Lists/isListPalindrome.php
// Singly-linked lists are already defined with this interface:
// class ListNode {
//   public $value;
//   public $next;
//
//   public function __construct($x) {
//     $this->value = $x;
//     $this->next = null;
//   }
// }
//
function solution($list) {
    //get the middle and reverse the first half
    $middle=$fast=$list;
    $reverse=null;
    while($fast && $fast->next!=null){
        $middle = $middle->next;
        $fast = $fast->next->next;
        
        $temp = $list->next;
        $list->next = $reverse;
        $reverse = $list;
        $list = $temp;
    }
    //fast left 1, odd number of nodes
    if($fast){
        $middle=$middle->next;
    }
    
    while ($middle) {
        if ($reverse->value != $middle->value) return false;
        $middle = $middle->next;
        $reverse = $reverse->next;
    }
    return true;
}
